rtrlib/aspa/aspa.c

```diff
@@ -176,368 +268,581 @@ static int compare_asns(const void *a, const void *b)
 			pthread_rwlock_unlock(&aspa_table->lock);
 			return ASPA_ERROR;
 		}
+		pthread_rwlock_unlock(&aspa_table->lock);
 	}
 
-	uint32_t **unused_sets = NULL;
-	size_t unused_size = 0;
-
-#ifdef ASPA_UPDATE_IN_PLACE
-	enum aspa_status res = aspa_table_update_internal(aspa_table, rtr_socket, array, operations, len, revert,
-							  failed_operation, &unused_sets, &unused_size);
-
-	pthread_rwlock_unlock(&aspa_table->lock);
-#else
-	pthread_rwlock_unlock(&aspa_table->lock);
+	assert(node);
+	assert(*node);
+	assert((*node)->aspa_array);
 
+	// Create new array that will hold updated record data
 	struct aspa_array *new_array = NULL;
 	if (aspa_array_create(&new_array) != ASPA_SUCCESS) {
 		return ASPA_ERROR;
 	}
 
-	enum aspa_status res = aspa_table_update_internal(aspa_table, rtr_socket, array, new_array, operations, len,
-							  failed_operation, &unused_sets, &unused_size);
-#endif
+	// Populate new_array
+	pthread_rwlock_rdlock(&aspa_table->lock);
+	enum aspa_status res = aspa_table_compute_update_internal(rtr_socket, (*node)->aspa_array, new_array,
+								  operations, count, &(*update)->failed_operation);
+	pthread_rwlock_unlock(&aspa_table->lock);
+
+	if (res == ASPA_SUCCESS) {
+		(*update)->node = *node;
+		(*update)->new_array = new_array;
+	} else {
+		(*update)->node = NULL;
+		(*update)->new_array = NULL;
 
-	if (finalization_args && !*finalization_args) {
-		*finalization_args = lrtr_malloc(sizeof(struct aspa_update_finalization_args));
+		// Update computation failed so release newly created array.
+		// We must not release associated provider arrays here.
+		aspa_array_free(new_array, false);
 	}
+	return res;
+}
+
+void aspa_table_apply_update(struct aspa_update *update)
+{
+	// Not going to re-apply update or apply if computation failed
+	if (!update || !update->table || !update->operations || !update->node || !update->new_array ||
+	    !!update->failed_operation) {
+		return;
+	}
+
+	struct aspa_array *old_array = update->node->aspa_array;
+
+	pthread_rwlock_wrlock(&update->table->lock);
+	update->node->aspa_array = update->new_array;
+	pthread_rwlock_unlock(&update->table->lock);
+
+	// Free old array
+	aspa_array_free(old_array, false);
+	old_array = NULL;
 
-	if (finalization_args && *finalization_args) {
-#ifndef ASPA_UPDATE_IN_PLACE
-		if (res == ASPA_SUCCESS) {
-			(*finalization_args)->old_array = &(*node)->aspa_array;
-			(*finalization_args)->new_array = new_array;
-		} else {
+	// Notify clients
+	for (size_t i = 0; i < update->operation_count; i++) {
+		struct aspa_update_operation *op = &update->operations[i];
+
+// Notify clients
+// We can directly use the operation array as the source for the diff
+// so we don't need to rely on first notifying clients about all records
+// in the old_array being removed and then every record in the new_array
+// being added again.
+#ifdef ASPA_NOTIFY_NO_OPS
+		aspa_table_notify_clients(update->table, &op->record, update->node->rtr_socket, op->type);
```

Comment:
maybe put macro around the if statement, because lines 509 and 512 are the same

Reply:
I've actually done that everywhere before, but it makes the code a lot more unreadable. And because that's all preprocessed I don't think that should be an issue